style.md

```diff
@@ -1515,6 +1516,114 @@ Note that the compiler will not generate errors when using predeclared
 identifiers, but tools such as `go vet` should correctly point out these and
 other cases of shadowing.
 
+### Avoid `init()`
+
+Avoid `init()` where possible. At a minimum, all code within `init()` should:
+
+1. Be completely deterministic, regardless of program environment or invocation.
```

Comment:
I think deterministic makes a lot of sense, although I'm not sure about environment. E.g., parsing environment variables to set up some global?

often done for things like gRPC log level control via environment variable, or optional overrides to a default hostname / port based on env vars, etc.

Reply:
I don't think there's anything wrong with using environment variables, I think the question is whether or not we recommend avoiding environment variables _in `init()`_. Why do those things need to happen in `init()`? Why not as a default config-providing function, or as part of a ctor, or as part of config loading, etc?

In your example, the problem that we quickly run into is an (ostensible) dichotomy between "provide explicit configuration paths through exported types" and "provide implicit global configuration". This is complicated further by the fact that, if you do some "init-level config", you're probably bifurcating your user-facing config (like gRPC does, which maybe isn't a model to follow).

What do you think?

Reply:
True, explicitly passing configuration is a lot cleaner. I think `init` is often used to configure globals or singletons which might not have an initializer.

That said, it's not a good pattern, so discouraging it makes sense to me.